### PR TITLE
i2263 - actually make it work

### PR DESCRIPTION
**Pull request by hillalex**
To run locally: `./scripts/run` and then follow instructions.

This assumes the cache file is in the format of the included `cache-example` file.

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,40 +1,27 @@
...
 
 app = Flask(__name__)
 
-
-def get_db_name():
-    return os.environ['BARMAN_DATABASE_NAME']
+config_path = "/etc/cm/config.json"
 
 
-def seconds_elapsed_since(timestamp):
```
Comment (hillalex) [inline app/main.py:18]: moved into submodule

File: app/main.py [hunk 2]
```diff
@@ -45,41 +34,25 @@ def barman_output_as_dict(text):
...
-        "barman_available_backups": status_values["No. of available backups"],
-        "barman_time_since_last_backup_seconds": since_last_backup,
-        "barman_time_since_last_backup_minutes":
-            since_last_backup / 60 if since_last_backup is not None else None,
-        "barman_time_since_last_backup_hours":
-            since_last_backup / 3600 if since_last_backup is not None else None,
-        "barman_time_since_last_backup_days":
-            since_last_backup / (3600 * 24) if since_last_backup is not None else None
-    }
+    if since_last_backup > 60 * 10:
```
Comment (richfitz) [inline app/main.py]: This should be configurable, in `config.json` perhaps as `max_age_seconds`?

File: app/main.py [hunk 3]
```diff
@@ -45,41 +34,25 @@ def barman_output_as_dict(text):
...
+        return {"responding": False}
+    else:
+        return status_values
```
Comment (richfitz) [inline app/main.py]: Do we need to return `responding: True` here in addition to the data?

File: app/main.py [hunk 4]
```diff
@@ -45,41 +34,25 @@ def barman_output_as_dict(text):
...
-        ms = label_metrics(ms, {"database": db_name})
-        return render_metrics(ms)
+        status = get_status()
+        ms = parse_status(status)
     except:
-        return render_metrics({
-            "barman_responding": False
-        })
+        ms = {"responding": False}
+    labels = get_labels()
```
Comment (richfitz) [inline app/main.py]: I forget how flask apps work but is it worth doing this on app startup?

**Review by richfitz — CHANGES_REQUESTED** [review]

**Comment by hillalex** [comment]
Note the more complex directory structure, with `app/app` is taken directly from the documentation of the docker image we're using: https://github.com/tiangolo/uwsgi-nginx-flask-docker#quickstart-for-bigger-projects-structured-as-a-python-package

As soon as I started adding more than one file into the app directory I ran into all sorts of problems with python module imports. Following the instructions in that README I was able to get it working both inside the main container and in the test container.

File: scripts/run [hunk 5]
```diff
@@ -1,9 +1,22 @@
...
+    echo "Please create a config.json file in the same format as the config-example.json file in this repo."
+    exit -1;
+fi
+
+HERE=${BASH_SOURCE%/*}
+$HERE/extract-config
+
+source config.sh
+
+docker build --tag $tag .
```
Comment (richfitz) [inline scripts/run]: Can we build the image on teamcity like the other images, push out to docker hub and our local registry? Or we can move this into a future ticket...

File: cache-example [hunk 6]
```diff
@@ -0,0 +1,2 @@
+metrics_created_at: 1539089746
```
Comment (richfitz) [inline cache-example]: So based on our slack discussions, this becomes:

```json
{
  "utc_seconds":  1539089746,
  "utc_string": "ignored",
  "data": {
    "some_metrics": 32178
  }
}
```

File: tests/test_main.py [hunk 7]
```diff
@@ -1,67 +1,56 @@
...
-def test_endpoint_handles_arbitrary_error(monkeypatch):
-    monkeypatch.setenv('BARMAN_DATABASE_NAME', 'test')
-    # the call to barman status and barman check will fail in this testing context
-    # causing an error
+from datetime import datetime, timedelta
+from app.app.main import parse_status, metrics
+import app
+
+
+def create_mock_data(timestamp):
```
Comment (richfitz) [inline tests/test_main.py:6]: if you update the example above, similar update here too

File: README.md [hunk 8]
```diff
@@ -1,12 +1,23 @@
...
-./run
+sudo -H pip3 install -r ./montagu_metrics/requirements.txt
+./scripts/run
 ```
+
+## Tests
+
+To run tests inside a docker container, with `config-example.json` mounted as the config file:
+```
+./scripts/teamcity.sh
```
Comment (richfitz) [inline README.md]: The teamcity build built for me a different container than the one that `run` built - is that expected?

**Review by richfitz — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (hillalex) [reply README.md]: yes, `run` doesn't copy tests into the container

**Comment by hillalex** [comment]
Tested and working with local barman running from this branch: `https://github.com/vimc/montagu-db/pull/108`

File: app/app/main.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+
+@app.route('/metrics')
+def metrics():
+    try:
+        status = get_status(settings.cache_file)
+        ms = parse_status(status, settings.max_age_seconds)
+    except Exception as e:
+        print(e)
+        ms = {"responding": False}
```
Comment (richfitz) [inline app/app/main.py]: Should we add a `{responding: True}` to the successful branch of the `try` block?

File: scripts/run [hunk 10]
```diff
@@ -1,9 +1,24 @@
...
+    exit -1;
+fi
+
+HERE=${BASH_SOURCE%/*}
+$HERE/extract-config
+
+source config.sh
+
+echo "Reading from cache located at ${cache_volume}"
+
```
Comment (richfitz) [inline scripts/run:16]: pull the image first perhaps?

**Review by MartinEden — CHANGES_REQUESTED** [review]
I'm still trying to get my head around how this all fits together.

Could you add the following to the README? I think they would both be helpful in the future, and helpful for me to understand things now.

* Description of what the config file format is
* Instructions on how to run the app minimally - e.g. create a volume with a file containing some metrics, as if we'd run something like the barman cron job, and then how to point this app at that file

**Review by richfitz — CHANGES_REQUESTED** [review]

**Comment by hillalex** [comment]
Ok, have made the README more detailed @MartinEden

File: app/app/main.py (hunk 9 above)
Comment (hillalex) [reply app/app/main.py]: That is happening but in `parse_status`, so maybe it would be clearer to move that logic here?

Comment (hillalex) [reply app/app/main.py]: Ah, I see what's going on. We're using `responding: False` to indicate both that the metrics have error-red, and that the cache is stale. I'll split into 2 different metrics.

File: scripts/run [hunk 11]
```diff
@@ -1,9 +1,25 @@
...
 docker run --rm \
-    -p 5000:80 \
+    -p $port:80 \
     --detach \
-    --name=barman_metrics_1 \
-    barman_metrics
+    -v $cache_volume:/app/cache \
+    -v $PWD/config.json:/etc/cm/config.json \
+    --name=$name \
+    docker.montagu.dide.ic.ac.uk:5000/cached-metrics:$GIT_ID
```
Comment (MartinEden) [inline scripts/run:27]: It would be nice to echo at something like "Metrics are now being served at http://localhost:$port/metrics".

**Review by MartinEden — CHANGES_REQUESTED** [review]
Okay, I think I've got my head around this now, and I am happy with the code as it stands other than this one minor suggestion.

**Review by richfitz — APPROVED** [review]

**Review by MartinEden — APPROVED** [review]

**Pull request merged**